### ISSUE TITLE
Enabling installation to inactive slot on Pixel devices

### ISSUE THREAD

**Issue by capntrips** [topjohnwu/Magisk#4955]
Device: Pixel 5a and Pixel 6
Android version: 12
Magisk version name: Canary
Magisk version code: 23014

I built a new version of `bootctl` against the Pixel 5a source:

```bash
repo init -u https://android.googlesource.com/platform/manifest -b android-12.0.0_r10
repo sync -c -j8
. build/envsetup.sh
lunch aosp_barbet-userdebug
mmma system/extras/bootctl
```
With a small change to statically link the missing libraries:

```diff
diff --git a/bootctl/Android.bp b/bootctl/Android.bp
index f63871cf..7c60f65b 100644
--- a/bootctl/Android.bp
+++ b/bootctl/Android.bp
@@ -26,10 +26,14 @@ cc_binary {
         "-Werror",
     ],
 
-    shared_libs: [
+    static_libs: [
         "android.hardware.boot@1.0",
         "android.hardware.boot@1.1",
         "android.hardware.boot@1.2",
+    ],
+    shared_libs: [
+        "liblog",
+        "libcutils",
         "libhidlbase",
         "libutils",
     ],
```

I included it in a build of Magisk with installation to the inactive slot enabled for Pixel 5a and was able to take an OTA using the method for A/B devices from the [guide](https://topjohnwu.github.io/Magisk/ota.html#devices-with-ab-partitions).

An XDA user reported success with a similar build on a Pixel 6 (`android-12.0.0_r12`, `aosp_oriole-userdebug`).

Pixel 6 uses the `devinfo` partition for boot control, rather than `gpt`. See [here](https://android.googlesource.com/device/google/gs101/+/refs/tags/android-12.0.0_r12/interfaces/boot/1.2/BootControl.cpp#208) and [here](https://android.googlesource.com/device/google/gs101/+/refs/tags/android-12.0.0_r12/interfaces/boot/1.2/BootControl.cpp#228).

I suspect the newer build of `bootctl` will work for most devices before Pixel 6, though it is easy enough to [check](https://android.googlesource.com/device/google/gs101/+/refs/tags/android-12.0.0_r12/interfaces/boot/1.2/BootControl.cpp#147) for a valid `devinfo` partition.

I'm happy to submit PRs adding support, but I figured I had better get some feedback before I do the work.

**Comment by vvb2060** [comment]
Why Pixel6 can't use bootctl of `android-12.0.0_r12` tag? Does `isDevInfoValid()` return `false`?

**Comment by capntrips** [comment]
The binary works fine and makes the changes, but my Pixel 6 bootlooped when I tried to use that binary in [`post_ota`](https://github.com/topjohnwu/Magisk/blob/1e05f8c6461594959e897f955a043a7f29f21c4f/app/src/main/res/raw/manager.sh#L93).

I planning to test a modified `bootctl` that uses [`getActiveBootSlot`](https://android.googlesource.com/device/google/gs101/+/b381ac1d0ccd8f1aaefcf4d21de18d7837465346/interfaces/boot/1.2/BootControl.cpp#397) instead of [`getCurrentSlot`](https://android.googlesource.com/device/google/gs101/+/b381ac1d0ccd8f1aaefcf4d21de18d7837465346/interfaces/boot/1.2/BootControl.cpp#192) in [`markBootSuccessful`](https://android.googlesource.com/device/google/gs101/+/b381ac1d0ccd8f1aaefcf4d21de18d7837465346/interfaces/boot/1.2/BootControl.cpp#207) when the December update is released.

**Comment by capntrips** [comment]
After helping someone else through the process, it's possible that I pressed the restart button within the updater during my own testing, so an updated stock `bootctl` may work, if the device is restarted outside of it.

**Comment by capntrips** [comment]
An XDA user reported success taking the OTA using a build of Magisk that enabled installation to the inactive slot on a Pixel 6 and included an updated build of `bootctl`.

On a related note, I was able to make a working build of `bootctl` with NDK 23.0.7599858, if you're interested in including the `bootctl` build as a part of the Magisk build process. Unfortunately, it required the sources for `jsoncpp`, `libbacktrace`, `libbase`, `libcutils`, `libcxx`, `libcxxabi`, `libdl`, `libfmq`, `libhidl`, `libhwbinder`, `liblog`, `libprocessgroup`, `libsystem`, `libutils`, and `libvndksupport`, plus the output of `hidl-gen` for `android.hardware.boot@1.0`, `android.hardware.boot@1.1`, `android.hardware.boot@1.2`, `android.hidl.base@1.0`, `android.hidl.manager@1.0`, `android.hidl.manager@1.1`, and `android.hidl.manager@1.2`.

Someone with a better understanding of how linking shared objects works might be able to simplify it.

**Comment by pndwal** [comment]
Is this closed in #5134 now?

**Comment by capntrips** [comment]
9c93fe6 updates `bootctl`, but it does not enable installation to the invactive slot on Pixel devices.
